gnc/sub8_perception/src/sub8_pcl/sub8_gateway.py

```diff
@@ -0,0 +1,202 @@
+    print("vectors " + str(len(vectors)))
+    # find vectors that correspond to edges
+
+    # check for pairs that share a point
+    vectors2 = [item for item in vectors]
+    vector_pairs = []
+    for vector in vectors:
+        for vector2 in vectors2:
+            if (vector.point_set[1] == vector2.point_set[0]):
+                if not(vector.equals(vector2)):
+                    vector_pairs.append((vector, vector2))
+    print("pairs " + str(len(vector_pairs)))
+    # check for perpendicular pairs
+    temp = []
+    for pair in vector_pairs:
+        if Vector.is_ortho(pair[0], pair[1], tolerance):
+            temp.append(pair)
+    vector_pairs = temp
+    print("pairs " + str(len(vector_pairs)))
+    # check for overlapping pairs
+    candidates = []
+    for pair1 in vector_pairs:
+        for pair2 in vector_pairs:
+            if not((pair1[0].equals(pair2[0]) and pair1[1].equals(pair2[1])) or (pair1[0].equals(pair2[1]) and pair1[0].equals(pair2[1]))):  # they are not equal
+                if pair1[0].equals(pair2[0]) and (abs(float(pair1[1].magnitude - pair2[1].magnitude) / pair1[1].magnitude) < 0.1):
+                    candidates.append((pair1, pair2))
+                    continue
+    # check for correct shape
+    final_results = []
+    for item in candidates:
+        points = []
+        for pair in item:
+            for vector in pair:
+                for point in vector.point_set:
+                    points.append(point)
+        c_values = [point[1] for point in points]
+        r_values = [point[0] for point in points]
+        r_min, r_max = min(r_values), max(r_values)
+        c_min, c_max = min(c_values), max(c_values)
+        top_left = (r_min, c_min)
+        dimensions = (r_max-r_min, c_max-c_min)
+        final_results.append((top_left, dimensions))
+    # select a candidate
+    if len(final_results) > 0:
+        final_results = final_results[0]
+    return final_results
+
+def test_img(img):
+    hsv_img = color_filter(img, target_hue=12, t_hue=200)
+    #edges = edge_map(hsv_img)
+    edges = []
+    gray_img = cv2.cvtColor(hsv_img, cv2.COLOR_BGR2GRAY)
+    corners = corner_map(gray_img)
+    #plt.gray()
+    #plt.imshow(gray_img)
+    #plt.show()
+    centroids = find_centroids(corners)
+    #plt.plot(centroids)
+    results = find_shape(centroids, edges)
+    #plt.imshow(edges)
+    #plt.show()
+    return json.dumps(results)
+
+############### COMMUNICATION ################
+data_buffer = None  # allows communication between publisher and subscriber
+
+def process_img(img):
+    global data_buffer
+    bridge = cv_bridge.CvBridge()
+    img = bridge.imgmsg_to_cv2(img)  # converts image to a format compatible with OpenCV
+    img = np.uint8(img)  # keeps OpenCV happy
+    data_buffer = test_img(img)
+
+def start_gateway_detector():
+    global data_buffer
+    # start listener
+    rospy.init_node('gateway_detector')
+    rospy.Subscriber(camera, ROS_img, callback=process_img)

```

Comment:
To avoid stress, you can use an [image subscriber](https://github.com/uf-mil/Sub8/blob/master/utils/sub8_ros_tools/sub8_ros_tools/image_helpers.py)